tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source.ts

```diff
@@ -40,62 +35,39 @@ export class QueryParamsFeatureFlagDataSource
     // Set feature flag value for query parameters that are explicitly
     // specified. Feature flags for unspecified query parameters remain unset so
     // their values in the underlying state are not inadvertently changed.
-    const featureFlags: Partial<FeatureFlags> = enableMediaQuery
-      ? this.getPartialFeaturesFromMediaQuery()
-      : {};
-    if (params.has(EXPERIMENTAL_PLUGIN_QUERY_PARAM_KEY)) {
-      featureFlags.enabledExperimentalPlugins = params.getAll(
-        EXPERIMENTAL_PLUGIN_QUERY_PARAM_KEY
-      );
-    }
-    if (params.has('tensorboardColab')) {
-      featureFlags.inColab = params.get('tensorboardColab') === 'true';
-    }
-    if (params.has(SCALARS_BATCH_SIZE_PARAM_KEY)) {
-      featureFlags.scalarsBatchSize = Number(
-        params.get(SCALARS_BATCH_SIZE_PARAM_KEY)
-      );
-    }
-
-    if (params.has(ENABLE_COLOR_GROUP_QUERY_PARAM_KEY)) {
-      featureFlags.enabledColorGroup =
-        params.get(ENABLE_COLOR_GROUP_QUERY_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_COLOR_GROUP_BY_REGEX_QUERY_PARAM_KEY)) {
-      featureFlags.enabledColorGroupByRegex =
-        params.get(ENABLE_COLOR_GROUP_BY_REGEX_QUERY_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_DARK_MODE_QUERY_PARAM_KEY)) {
-      featureFlags.defaultEnableDarkMode =
-        params.get(ENABLE_DARK_MODE_QUERY_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_LINKED_TIME_PARAM_KEY)) {
-      featureFlags.enabledLinkedTime =
-        params.get(ENABLE_LINKED_TIME_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_CARD_WIDTH_SETTING_PARAM_KEY)) {
-      featureFlags.enabledCardWidthSetting =
-        params.get(ENABLE_CARD_WIDTH_SETTING_PARAM_KEY) !== 'false';
-    }
+    const featureFlags: Partial<Record<keyof FeatureFlags, FeatureFlagType>> =
+      enableMediaQuery ? this.getPartialFeaturesFromMediaQuery() : {};
+    Object.entries(FeatureFlagQueryParameters).forEach(
+      ([flagName, flagMetadata]) => {
+        const featureValue = this.getFeatureValue(flagMetadata);
+        if (featureValue !== null) {
+          const f = flagName as keyof FeatureFlags;
+          featureFlags[f] = featureValue;
+        }
+      }
+    );
+    return featureFlags as Partial<FeatureFlags>;
+  }
 
-    if (params.has(FORCE_SVG_RENDERER)) {
-      featureFlags.forceSvg = params.get(FORCE_SVG_RENDERER) !== 'false';
+  protected getFeatureValue(
+    flagMetadata: FeatureFlagMetadata
+  ): FeatureFlagType {
+    const params = this.queryParams.getParams();
+    const queryParamOverride = flagMetadata.queryParamOverride;
+    if (!queryParamOverride || !params.has(queryParamOverride)) {
+      return null;
     }
-
-    if (params.has(ENABLE_DATA_TABLE_PARAM_KEY)) {
-      featureFlags.enabledScalarDataTable =
-        params.get(ENABLE_DATA_TABLE_PARAM_KEY) !== 'false';
+    const paramValues: BaseFeatureFlagType[] = this.queryParams
+      .getParams()
+      .getAll(queryParamOverride)
```

Comment:
If a user has more than one instance of a query param it will always be parsed as an array. This can lead to broken states. Consider adding another optional property to the `FeatureFlagMetadata`

Reply:
Yeah... this is easy enough to fix that it seems worth fixing